drivers: eth: sam: Replace constants by devicetree values

nandojve opened the pull request:
The zephyr sam gmac driver don't get both register address and peripheral id from devicetree.  Drop headers constants in favor
of devicetree values.

This fix wrong Atmel SAME7x/SAMV7x gmac register address and add missing peripheral id property for SAM family.

The problem was found running tests for #34718 .

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
drivers: eth: sam: Replace constants by devicetree values

The zephyr sam gmac driver don't get register address and, in some
cases, peripheral id from devicetree.  This replace headers constants
in favor of devicetree values.

This fix wrong Atmel SAME7x/SAMV7x gmac register address and add
missing peripheral id property for SAM family.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

stephanosio commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -2218,14 +2218,14 @@ static const struct soc_gpio_pin pins_eth0[] = ATMEL_SAM_DT_INST_PINS(0);
 static const struct eth_sam_dev_cfg eth0_config = {
-	.regs = GMAC,
-	.periph_id = ID_GMAC,
+	.regs = (Gmac *) DT_INST_REG_ADDR(0),
```
Spurious space
```suggestion
	.regs = (Gmac *)DT_INST_REG_ADDR(0),
```

stephanosio commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -2218,14 +2218,14 @@ static const struct soc_gpio_pin pins_eth0[] = ATMEL_SAM_DT_INST_PINS(0);
 #endif
 	.config_func = eth0_irq_config,
-	.phy = {GMAC, CONFIG_ETH_SAM_GMAC_PHY_ADDR},
+	.phy = {(Gmac *) DT_INST_REG_ADDR(0), CONFIG_ETH_SAM_GMAC_PHY_ADDR},
```
Spurious space
```suggestion
	.phy = {(Gmac *)DT_INST_REG_ADDR(0), CONFIG_ETH_SAM_GMAC_PHY_ADDR},
```

stephanosio reviewed (dismissed):
Should update SAM E5x series as well:
https://github.com/zephyrproject-rtos/zephyr/blob/395b0c3a2469e9e80077dc9aec4fe88ecfb11baf/dts/arm/atmel/same5x.dtsi#L11

stephanosio commented on dts/bindings/ethernet/atmel,sam-gmac.yaml:
```diff
@@ -41,6 +41,11 @@ properties:
 
+    peripheral-id:
+      type: int
+      description: Peripheral ID
```
would be nice to note here that `peripheral-id` is only applicable to the SAM (and not SAM0) family devices.

mnkp replied:
SAM0 is a different SoC family. Nothing in this file is applicable to SAM0 devices, not only `peripheral-id` property. I would find such comment confusing.

stephanosio replied:
@mnkp GMAC is available in SAM E5x, which is SAM0 family.

mnkp replied:
Sorry, I missed this. @stephanosio Your comment is valid.

nandojve replied:
Add a default value to peripheral-id. With that, all SAM0 families will be OK. Besides peripheral-id be mandatory for any SAM, I think this is best relation between devicetree standards vs shared functionality. Otherwise, it will be odd have required: true and a default: 0 or peripheral-id at SAM0 dtsi.

mnkp commented on dts/bindings/ethernet/atmel,sam-gmac.yaml:
```diff
@@ -41,6 +41,11 @@ properties:
+    peripheral-id:
+      type: int
+      description: Peripheral ID
+      required: false
```
I believe this should be a required property, i.e. we need `true` here. Without the correct periph id value the driver will not work.

stephanosio replied:
That would make any SAM0 devices (e.g. SAM E5x) using this to specify dummy `peripheral-id` here.

mbolivar-nordic reviewed (changes requested):
Can you please split the whitespace changes into their own commit?

Also please choose one of the following styles:

```
description: foo
```

or

```
description: |
  foo
```

do not use

```
description:
  foo
```

nandojve committed:
drivers: eth: sam: Fix style and documentation

This fixes properties documentation and two devicetree styles:

 - properties with wrong code identation and
 - property description tag style

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

mbolivar-nordic commented on dts/bindings/ethernet/atmel,sam-gmac.yaml:
```diff
@@ -15,47 +15,47 @@ properties:
-          and 10240 (jumbo frames).
-        default: 1518
+      type: int
+      default: 1518
```
Thanks for splitting that up, it's easier to read now.

If you have time, please also add a justification for this default, since you are fixing DT bindings issues. But that was a problem before this PR so I'm not blocking on that request.

mbolivar-nordic commented on dts/bindings/ethernet/atmel,sam-gmac.yaml:
```diff
@@ -15,47 +15,47 @@ properties:
-          Specifies maximum speed in Mbit/s supported by the device.
-        default: 100
+      type: int
+      default: 100
```
Same here; why is this the default? Can you justify it if you have time? Again not blocking since the problem existed before this PR.

mbolivar-nordic commented on dts/bindings/ethernet/atmel,sam-gmac.yaml:
```diff
@@ -15,47 +15,47 @@ properties:
+      enum:
+        - "rmii"
+        - "mii"
+      default: "rmii"
```
Same

mbolivar-nordic commented on dts/bindings/ethernet/atmel,sam-gmac.yaml:
```diff
@@ -41,6 +41,16 @@ properties:
 
+    peripheral-id:
+      type: int
+      default: 0
```
This default is new and its value must be justified to follow the rules in the DT bindings documentation. This is a blocking change request.

mbolivar-nordic commented on dts/bindings/ethernet/atmel,sam-gmac.yaml:
```diff
@@ -41,6 +41,16 @@ properties:
+        A peripheral identifier is required for the control of the peripheral
+        interrupt with the Nested Vectored Interrupt Controller (NVIC) and
+        control of the peripheral clock with the Power Management Controller
+        (PMC).  This parameter is mandatory for any SAM family.  SAM0 family
+        uses another strategy to define both interrupt and clock control.
```
I'm confused about this. It says it's mandatory but it has a default value, which doesn't make sense.

mbolivar-nordic reviewed (changes requested):
Thanks for the update

mbolivar-nordic commented on dts/bindings/ethernet/atmel,sam-gmac.yaml:
```diff
@@ -55,6 +55,21 @@ properties:
+        associated with a specific peripheral.  This clock line is defined at
+        Power Management Controller (PMC) and it enables the peripheral.
+
+        The SAM0 family uses another strategy to define peripheral clock.
+        To avoid an odd property at gmac devicetree node for SAM0 family a
+        default value was defined.
```
Thanks for explaining this better.

If the hardware behaves differently, I wonder why a different compatible is not used instead? Can you please justify this decision?

nandojve replied:
Thank you for pushing on this. Can you check again, please?

nandojve committed:
dts: eth: sam: Split generic GMAC compatible

The current GMAC compatible not allow especialize properties by SoC
family.  Split current generic Atmel GMAC compatible into two new
compatibles which are defined by SoC families.  This increase the
freedom and avoid odd situations.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

mbolivar-nordic reviewed (approved):
Thanks, LGTM!

cfriedt commented:
@nandojve - are you able to address the comment made by @stephanosio, or is it no longer relevant?

nandojve commented:
> @nandojve - are you able to address the comment made by @stephanosio, or is it no longer relevant?

Hi @cfriedt ,

I already checked and current GMAC address is correct for SAM E5x. No changes are required.
Thank you for double check!
